### PR TITLE
Prevent non-data nodes to start with dangling indices

### LINKED ISSUE #27073: Dangling indices living in non-data nodes are detected and auto-imported
**Elasticsearch version** (`bin/elasticsearch --version`):
```
Version: 5.5.3, Build: 9305a5e/2017-09-07T15:56:59.599Z, JVM: 1.8.0_151
```

**Plugins installed**: []

**JVM version** (`java -version`): 
```
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)
```

**OS version** (`uname -a` if on a Unix-like system): `Darwin Thiagos-MacBook-Pro.local 17.0.0 Darwin Kernel Version 17.0.0: Thu Aug 24 21:48:19 PDT 2017; root:xnu-4570.1.46~2/RELEASE_X86_64 x86_64`

**Description of the problem including expected versus actual behavior**:

If a non-data node, that contains dangling indices in it's data path, joins a cluster these dangling indices will be detected and auto-imported.

IMO, a non-data node that contains index data in it's data path is probably accidental and unintended. In this case, those dangling indices should not be detected, better yet if the node does not even starts (maybe a bootstrap check that fails if a non-data node contains index data in it's data path).

**Steps to reproduce**:

This can be done in a single machine:
 1. Start `node-1` with `bin/elasticsearch -E path.data=/Users/thiago/data-1 -E node.name=node-1`
 2. Start `node-2` with `bin/elasticsearch -E path.data=/Users/thiago/data-2 -E node.name=node-2`
 3. Create an index `test` configured with `1S/0R` with `curl -XPUT localhost:9200/test -d '{ "settings": { "index": { "number_of_shards": 1, "number_of_replicas": 0 }  } }' -H "Content-Type: application/json"`
 4. Create a document `curl -XPOST localhost:9200/test -d '{ "test": 1 }' -H "Content-Type: application/json"`
 5. Stop both nodes
 6. Check which data directory, either `data-1` or `data-2`, that the shard for index `test` was created in and delete the _other_ empty data directory (so we effectively make a dangling index). 
 7. Consider that `data-2` was deleted. So start `node-2` again with `bin/elasticsearch -E path.data=/Users/thiago/data-2 -E node.name=node-2`
 8. Start `node-1` (which contains dangling indices) as a non-data node with `bin/elasticsearch -E path.data=/Users/thiago/data-1 -E node.name=node-1 -E node.data=false`

**Provide logs (if relevant)**:

After non-data node `node-1` starts, `node-2` will detect and auto-import dangling indices even though `node-1` is a non-data node: 
```
[2017-10-21T18:02:14,158][INFO ][o.e.g.LocalAllocateDangledIndices] [node-2] auto importing dangled indices [[test/R2Nh9sERThmkJ-0IZ0ppwA]/OPEN] from [{node-1}{RqWMW2AeSXWOpkUm4cT1TA}{lEqpWLIhRqqU_n1DSFuv2Q}{127.0.0.1}{127.0.0.1:9301}]
```

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
Non data node (like pure master or pure coordinator node) could behaves like a data node in terms of picking up dangling indices for other data nodes will automatically import them. 

This PR forbids non data nodes to start if any data shard folders in data path are found (presence of `_state` metadata folders are allowed for master node)

Closes #27073

### CODE REVIEW

**Comment by vladimirdolzhenko** [comment]
@elasticmachine run gradle build tests

**Review by ywelsch — CHANGES_REQUESTED** [review]
I've left a few comments. Note that your check is never called on coordinating-only nodes.

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 1]
```diff
@@ -313,6 +314,33 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
...
+                                        danglingShards.add(path.toAbsolutePath());
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+            if (danglingShards.isEmpty() == false) {
+                throw new IllegalStateException("Non data node cannot have dangling indices,"
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: addressed in 1478209

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 2]
```diff
@@ -294,6 +294,7 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
             this.locks = nodeLock.locks;
             this.nodePaths = nodeLock.nodePaths;
             this.nodeLockId = nodeLock.nodeId;
+            checkDanglingIndices(settings);
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: addressed in 1478209

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 3]
```diff
@@ -313,6 +314,33 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
...
+            final List<Path> danglingShards = new ArrayList<>();
+            for (final NodePath nodePath : nodePaths) {
+                // path to shard is: <node-id>/indices/<index-uuid>/<shard-id>
+                if (Files.exists(nodePath.indicesPath) && Files.isDirectory(nodePath.indicesPath)) {
+                    try (DirectoryStream<Path> stream = Files.newDirectoryStream(nodePath.indicesPath)) {
+                        for (final Path index : stream) {
+                            try (DirectoryStream<Path> indexStream = Files.newDirectoryStream(index)) {
+                                for (Path path: indexStream) {
+                                    // it's ok for non data node to have _state folder
+                                    if (MetaDataStateFormat.STATE_DIR_NAME.equals(path.getFileName().toString()) == false) {
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: addressed in 1478209

File: server/src/main/java/org/elasticsearch/env/NodeEnvironment.java [hunk 4]
```diff
@@ -313,6 +314,33 @@ public NodeEnvironment(Settings settings, Environment environment) throws IOExce
         }
     }
 
+    private void checkDanglingIndices(Settings settings) throws IOException {
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/main/java/org/elasticsearch/env/NodeEnvironment.java]: addressed in 1478209

File: server/src/test/java/org/elasticsearch/gateway/DanglingIndicesIT.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.startsWith;
+
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class DanglingIndicesIT extends ESIntegTestCase {
+
+    public void testDanglingIndices() throws Exception {
+        // replicating the case described in https://github.com/elastic/elasticsearch/issues/27073
```
[the remark this answers is not in the document]
Comment (vladimirdolzhenko) [reply server/src/test/java/org/elasticsearch/gateway/DanglingIndicesIT.java]: I've extended at 47d1382 this test to make a clean up of unexpected directories (picking those up from error message) and start node again.

**Comment by vladimirdolzhenko** [comment]
@elasticmachine run gradle build tests

**Comment by vladimirdolzhenko** [comment]
@ywelsch I addressed your comments, could you pls have another look?

**Comment by vladimirdolzhenko** [comment]
@elasticmachine run gradle build tests

**Comment by ywelsch** [comment]
superseded by #37347 and #37748

**Pull request closed without merging**